runtime/buffer.h

```diff
@@ -339,24 +347,29 @@ void for_each_slice(std::size_t slice_rank, raw_buffer& buf, const F& f) {
     return;
   }
 
-  buf.rank -= 1;
-  void* old_base = buf.base;
-  if (dim.fold_factor() == dim::unfolded) {
-    index_t stride = dim.stride();
-    for (index_t i = min; i <= max; ++i, buf.base = offset_bytes(buf.base, stride)) {
-      for_each_slice(slice_rank, buf, f);
-    }
-  } else {
-    // Extent 1 dimensions are likely very common here. We can handle that case more efficiently first because the
-    // base already points to the min.
-    for_each_slice(slice_rank, buf, f);
-    for (index_t i = min + 1; i <= max; ++i) {
-      buf.base = offset_bytes(old_base, dim.flat_offset_bytes(i));
-      for_each_slice(slice_rank, buf, f);
+  std::array<void*, N> old_bases;
+  for (std::size_t n = 0; n < N; ++n) {
+    old_bases[n] = bufs[n].base;
+    bufs[n].rank -= 1;
+  }
+
+  // TODO: If performance is an issue here, we can make a plan (like `for_each_contiguous_slice` does) that avoids
+  // folding overhead.
+  index_t stride = dim.stride();
+  assert(dim.min() / dim.fold_factor() == dim.max() / dim.fold_factor());
```

Comment:
do these need to be `euclidean_div`?

Reply:
I was working on making a helper for this, when I realized I think these asserts aren't even right with `euclidean_div`. Let's leave this for now because it's the same assert used in a bunch of other places, and I'll look at fixing this for all of them.